pkg/observable/channel/replay.go

```diff
@@ -0,0 +1,109 @@
+package channel
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"pocket/pkg/observable"
+)
+
+const replayNotificationTimeout = 1 * time.Second
+
+var _ observable.Observable[any] = &replayObservable[any]{}
+
+type replayObservable[V any] struct {
+	*channelObservable[V]
+	size              int
+	notificationsMu   sync.RWMutex
+	notifications     []V
+	replayObserversMu sync.RWMutex
+	replayObservers   []observable.Observer[V]
+}
+
+// Replay returns an observable which replays the last n values published to the
+// source observable to new observers, before publishing new values.
+func Replay[V any](
+	ctx context.Context, n int,
+	srcObsvbl observable.Observable[V],
+) observable.Observable[V] {
+	// TODO_HACK/TODO_IMPROVE: more effort is required to make a generic replay
+	// observable; however, as we only have the one observable package (channel),
+	// and aren't anticipating need another, we can get away with this for now.
+	chanObsvbl, ok := srcObsvbl.(*channelObservable[V])
+	if !ok {
+		panic("Replay only supports channelObservable")
+	}
+
+	replayObsvbl := &replayObservable[V]{
+		channelObservable: chanObsvbl,
+		size:              n,
+		notifications:     make([]V, 0, n),
+	}
+
+	srcObserver := srcObsvbl.Subscribe(ctx)
+	go replayObsvbl.goBufferReplayNotifications(srcObserver)
+
+	return replayObsvbl
+}
+
+// Next synchronously returns the next value from the observable. This will always
+// return the first value in the replay buffer, if it exists.
+func (ro *replayObservable[V]) Next(ctx context.Context) V {
```

Comment:
I think what we'd need is a `Last(ctx, n)` which better serves the LastBlock & LastNBlock like scenarios. WDYT?

Reply:
Yes, thank you! Not sure why I couldn't see it for myself. :sweat_smile:

Reply:
I think the reason I went with `Next` was because we're using `replayObservable[V]` as an `Observable[V]` interface, which meant that whatever method we add had to be included in the interface. 

Thinking about it some more, I feel like we should:

1. Remove this method from `Observable[V]`
2. Add a `ReplayObservable[V]` interface type which embeds `Observable[V]` and includes `Last()`
3. Change the return type of `Replay()` to `ReplayObservable[V]` so that it's explicit everywhere that we're using a replay observable. This should improve readability.